### PR TITLE
feat: propagate VMExport labels and annotations to associated pod

### PR DESCRIPTION
**Pull request by chomatdam**
### What this PR does
**Before this PR:**
- A VM export CR and pod are created in the same namespace as the VM, not necessarily on the same network, or reachable by the kubelet doing the readiness probe.
- A VM export server deployed using `virtctl` has a hardcoded 2-minutes readiness timeout, not taking into account a new node may have to be scheduled for this pod

**After this PR:**
- End user can label or annotate the VM export object or ask for it through `virtctl vmexport` arguments, the generated export server will inherit from VM export labels and annotations
- End user can define the VM Export readiness timeout
- Expected VM export labels and annotations are now unit tested

### Why we need it and why it was done in this way
- **Scenario 1:** End user can schedule the VM export object on a specific node through labels
- **Scenario 2:** End user can set a Multus (or CNI-related) annotation to get VM Export server on a specific network reachable by kubelet
- **Scenario 3:** A new node is being scheduled (e.g. on AWS, at least 90s to spin up an EC2), the existing VM export readiness timeout can now be configured to not hit the default value (2 minutes)

### Special notes for your reviewer
- Tested in a k8s cluster (virtctl and label/annotation propagation)
- Open to discussion on what would be relevant to update in the user guide (if needed)
   - updating virtctl download/create args mentioned as comments?
   - a use case using Multus network annotation on VM export object?

### Checklist

- [x] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required (no design change)
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [x] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [x] Community: Conversation started on Slack #kubevirt-dev [here](https://kubernetes.slack.com/archives/C0163DT0R8X/p1727784027682779)

### Release note
```release-note
Added labels, annotations to VM Export resources and configurable pod readiness timeout
```

### CODE REVIEW

File: pkg/safepath/BUILD.bazel [hunk 1]
```diff
@@ -4,6 +4,7 @@ go_library(
     name = "go_default_library",
     srcs = [
         "safepath.go",
+        "safepath_darwin.go",
```
Comment (awels) [inline pkg/safepath/BUILD.bazel:7]: So I am sure this is useful, but it seems out of scope for this PR. It is not actually used anywhere else in this PR. Could you make a separate PR that contains this change and the reasoning why it should go in?

**Review by awels — CHANGES_REQUESTED** [review]
So I think the safepath stuff got in by accident, can you move it to a separate PR?

File: pkg/safepath/BUILD.bazel (hunk 1 above)
Comment (chomatdam) [reply pkg/safepath/BUILD.bazel:7]: Removing it, it was just to be able to run unit tests locally 👍 (on macOS)

**Review by awels — APPROVED** [review]
Looks good.
@mhenriks could you approve? It is allowing annotations/labels to get passed from the VMExport resource to the service and pods.

**Comment by mhenriks** [comment]
Thanks for the PR @chomatdam!

@0xFelix what do you think of the virtctl changes?

File: pkg/virtctl/vmexport/vmexport.go [hunk 2]
```diff
@@ -725,7 +762,7 @@ func GetManifestUrlsFromVirtualMachineExport(vmexport *exportv1.VirtualMachineEx
 
 // waitForVirtualMachineExport waits for the VirtualMachineExport status and external links to be ready
 func waitForVirtualMachineExport(client kubecli.KubevirtClient, vmeInfo *VMExportInfo, interval, timeout time.Duration) error {
-	err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+	err := wait.PollUntilContextTimeout(context.Background(), interval, timeout, false, func(ctx context.Context) (bool, error) {
```
Comment (0xFelix) [inline pkg/virtctl/vmexport/vmexport.go:765]: Previously this started polling immeditately, now it does not (4th arg set to `false`)

File: pkg/virtctl/vmexport/vmexport.go [hunk 3]
```diff
@@ -1082,7 +1125,7 @@ func translateServicePortToTargetPort(localPort string, remotePort string, svc k
 func waitForExportServiceToBeReady(client kubecli.KubevirtClient, vmeInfo *VMExportInfo, interval, timeout time.Duration) (*k8sv1.Service, error) {
 	service := &k8sv1.Service{}
 	serviceName := fmt.Sprintf("virt-export-%s", vmeInfo.Name)
-	err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+	err := wait.PollUntilContextTimeout(context.Background(), interval, timeout, false, func(ctx context.Context) (bool, error) {
```
Comment (0xFelix) [inline pkg/virtctl/vmexport/vmexport.go:1128]: Same as above

**Review by 0xFelix — COMMENTED** [review]
@mhenriks Changes look good so far. 

@chomatdam Can you add tests for the changes? Preferrably based on https://github.com/kubevirt/kubevirt/pull/12939 which should be merged soon.

File: pkg/virtctl/vmexport/vmexport_test.go [hunk 4]
```diff
@@ -254,6 +254,27 @@ var _ = Describe("vmexport", func() {
 			Expect(err.Error()).Should(ContainSubstring(expectedError))
 		})
 
+		It("VirtualMachineExport download fails when readiness timeout", func() {
+			testInit(defaultHandler)
+			vme := utils.VMExportSpecPVC(vmexportName, metav1.NamespaceDefault, "test-pvc", secretName)
+			vme.Status = &exportv1.VirtualMachineExportStatus{
+				Phase: exportv1.Pending,
+			}
+			utils.HandleVMExportGet(vmExportClient, vme, vmexportName)
```
Comment (0xFelix) [inline pkg/virtctl/vmexport/vmexport_test.go:263]: Please base your changes on the PR I mentioned and try to make proper use of the fake client. The `utils.*` helpers will be dropped soon.

Comment (chomatdam) [reply pkg/virtctl/vmexport/vmexport_test.go:263]: Thanks for the extra context, are you asking that I rebase my PR against your fork? Doing so will combine both our changes, and your code not being merged into this repository yet, I can't update the PR to diff between your fork and mine right now
If your fork has to go first, I guess I can also wait until your changes are there? It will be minor test updates anyway

Comment (mhenriks) [reply pkg/virtctl/vmexport/vmexport_test.go:263]: Wait for #12939 to merge to main and rebase your branch on that. You could of course temporarily rebase on @0xFelix's branch though if that will help get the code written

File: pkg/virtctl/vmexport/vmexport_test.go [hunk 5]
```diff
@@ -267,6 +267,24 @@ var _ = Describe("vmexport", func() {
...
+
+			_, err := virtClient.ExportV1beta1().VirtualMachineExports(metav1.NamespaceDefault).Create(context.Background(), vme, metav1.CreateOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			err = runDownloadCmd(
+				setFlag(vmexport.PVC_FLAG, pvcName),
+				setFlag(vmexport.OUTPUT_FLAG, outputPath),
+				setFlag(vmexport.READINESS_TIMEOUT_FLAG, "1ns"),
+			)
+			expectedError := "context deadline exceeded"
```
Comment (0xFelix) [inline pkg/virtctl/vmexport/vmexport_test.go:284]: nit: it's okay to inline this

File: pkg/virtctl/vmexport/vmexport_test.go [hunk 6]
```diff
@@ -620,6 +638,23 @@ var _ = Describe("vmexport", func() {
 			Expect(err).ToNot(HaveOccurred())
 			Expect(*vme.Spec.TTLDuration).To(Equal(ttl))
 		})
+
+		It("Succesfully create VirtualMachineExport with custom labels and annotations", func() {
+			labels := map[string]string{"label-key": "label-value"}
+			annotations := map[string]string{"annotation-key": "annotation-value"}
```
Comment (0xFelix) [inline pkg/virtctl/vmexport/vmexport_test.go:644]:
```suggestion
			const (
					labelKey        = "label-key"
					labelValue      = "label-value"
					annotationKey   = "annotation-key"
					annotationValue = "annotation-key"
			)
```

File: pkg/virtctl/vmexport/vmexport_test.go [hunk 7]
```diff
@@ -620,6 +638,23 @@ var _ = Describe("vmexport", func() {
...
+
+			err := runCreateCmd(
+				setFlag(vmexport.PVC_FLAG, pvcName),
+				setFlag(vmexport.LABELS_FLAG, fmt.Sprintf("%s=%s", "label-key", "label-value")),
+				setFlag(vmexport.ANNOTATIONS_FLAG, fmt.Sprintf("%s=%s", "annotation-key", "annotation-value")),
```
Comment (0xFelix) [inline pkg/virtctl/vmexport/vmexport_test.go:649]:
```suggestion
				setFlag(vmexport.LABELS_FLAG, fmt.Sprintf("%s=%s", labelKey, labelValue)),
				setFlag(vmexport.ANNOTATIONS_FLAG, fmt.Sprintf("%s=%s", annotationKey, annotationValue)),
```

File: pkg/virtctl/vmexport/vmexport_test.go [hunk 8]
```diff
@@ -620,6 +638,23 @@ var _ = Describe("vmexport", func() {
...
+			)
+			Expect(err).ToNot(HaveOccurred())
+
+			vme, err := virtClient.ExportV1beta1().VirtualMachineExports(metav1.NamespaceDefault).Get(context.Background(), vme.Name, metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(vme.Labels).To(Equal(labels))
+			Expect(vme.Annotations).To(Equal(annotations))
```
Comment (0xFelix) [inline pkg/virtctl/vmexport/vmexport_test.go:656]:
```suggestion
			Expect(vme.Labels).To(HaveKeyWithValue(labelKey, labelValue))
			Expect(vme.Annotations).To(HaveKeyWithValue(annotationKey, annotationValue))
```

File: pkg/virtctl/vmexport/vmexport_test.go [hunk 9]
```diff
@@ -267,6 +267,23 @@ var _ = Describe("vmexport", func() {
...
+
+			_, err := virtClient.ExportV1beta1().VirtualMachineExports(metav1.NamespaceDefault).Create(context.Background(), vme, metav1.CreateOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			err = runDownloadCmd(
+				setFlag(vmexport.PVC_FLAG, pvcName),
+				setFlag(vmexport.OUTPUT_FLAG, outputPath),
+				setFlag(vmexport.READINESS_TIMEOUT_FLAG, "1ns"),
+			)
+			Expect(err.Error()).To(Equal("context deadline exceeded"))
```
Comment (0xFelix) [inline pkg/virtctl/vmexport/vmexport_test.go:284]:
```suggestion
			Expect(err).To(MatchError("context deadline exceeded"))
```

Comment (0xFelix) [reply pkg/virtctl/vmexport/vmexport_test.go:284]: One last nit

**Review by 0xFelix — COMMENTED** [review]
One last nit + could you please squash your commits?
Otherwise looks good to me.

Thanks for putting up with me! :)

**Comment by chomatdam** [comment]
> One last nit + could you please squash your commits? Otherwise looks good to me.
> 
> Thanks for putting up with me! :)

No problem @0xFelix, it's updated - I'm also on the Kubernetes Slack if needed if any other nits 🙂

**Review by 0xFelix — APPROVED** [review]
/lgtm

**Comment by 0xFelix** [comment]
Thanks!

**Comment by chomatdam** [comment]
@0xFelix FYI had to rebase and lost the `/lgtm`

**Comment by mhenriks** [comment]
@chomatdam build tree is not clean, try `make generate` and/or `make lint`

**Comment by chomatdam** [comment]
@mhenriks done, a VM export `BUILD.bazel` change for an uuid import + import reordered in the .go file, let me know if it fixes it

**Comment by mhenriks** [comment]
Two tests seem to be consistently failing

**Comment by chomatdam** [comment]
@mhenriks extra help to run it or fix it would be appreciated, I'm on aarch64
- `make bazel-test` works after a few "darwin" tweaks but is scoped to unit tests
- running `tests/storage/memorydump.go` apart, I'm getting some issues too ([output.log](https://github.com/user-attachments/files/17410567/output.log))

**Comment by mhenriks** [comment]
> @mhenriks extra help to run it or fix it would be appreciated, I'm on aarch64
> 
> * `make bazel-test` works after a few "darwin" tweaks but is scoped to unit tests
> * running `tests/storage/memorydump.go` apart, I'm getting some issues too ([output.log](https://github.com/user-attachments/files/17410567/output.log), tried with `CGO_ENABLED=1`)

@chomatdam I think the issue is that your newly added readinessTimeout field is not being set here: https://github.com/pelotech/kubevirt/blob/add-labels-annotations-to-vmexport/pkg/virtctl/memorydump/memorydump.go#L329-L343

**Comment by chomatdam** [comment]
thanks @mhenriks 🙃

**Comment by mhenriks** [comment]
> thanks @mhenriks 🙃

yeah that seems to have fixed the issue. Thanks for the PR!

/lgtm
/approve
